File: 3734-lexicographically-smallest-palindromic-permutation-greater-than-target/3734-lexicographically-smallest-palindromic-permutation-greater-than-target.cpp
class Solution {
public:
    string build(string &left, char mid) {
        string res = left;

        if (mid != '#')
            res += mid;

        for (int i = (int)left.size() - 1; i >= 0; i--)
            res += left[i];

        return res;
    }

    string lexPalindromicPermutation(string s, string target) {
        int n = s.size();

        vector<int> cnt(26, 0);

        for (char c : s)
            cnt[c - 'a']++;

        // Check if a palindrome can be formed
        int odd = 0;
        char mid = '#';

        for (int i = 0; i < 26; i++) {
            if (cnt[i] % 2) {
                odd++;
                mid = char('a' + i);
            }
        }

        if (odd > 1)
            return "";

        // Characters available for the left half
        vector<int> half(26);

        for (int i = 0; i < 26; i++)
            half[i] = cnt[i] / 2;

        int halfLen = n / 2;

        string left;

        /*
         * Build the answer from left to right.
         */
        for (int pos = 0; pos < halfLen; pos++) {

            bool found = false;

            for (int c = 0; c < 26; c++) {

                if (half[c] == 0)
                    continue;

                // Try character c
                half[c]--;
                left.push_back(char('a' + c));

                /*
                 * Put all remaining characters in descending order.
                 * This creates the largest palindrome having this prefix.
                 */
                string maxLeft = left;

                for (int x = 25; x >= 0; x--) {
                    maxLeft.append(
                        half[x],
                        char('a' + x)
                    );
                }

                string candidate = build(maxLeft, mid);

                /*
                 * If even the largest possible palindrome
                 * is not greater than target, this character
                 * cannot be used here.
                 */
                if (candidate > target) {
                    found = true;
                    break;
                }

                // Undo
                left.pop_back();
                half[c]++;
            }

            if (!found)
                return "";
        }

        string ans = build(left, mid);

        if (ans > target)
            return ans;

        return "";
    }
};

File: 3734-lexicographically-smallest-palindromic-permutation-greater-than-target/3734-lexicographically-smallest-palindromic-permutation-greater-than-target.js
/**
 * @param {string} s
 * @param {string} target
 * @return {string}
 */
var lexPalindromicPermutation = function(s, target) {

    const n = s.length;

    const cnt = new Array(26).fill(0);

    for (const ch of s) {
        cnt[ch.charCodeAt(0) - 97]++;
    }

    let odd = 0;
    let mid = '#';

    for (let i = 0; i < 26; i++) {
        if (cnt[i] % 2 === 1) {
            odd++;
            mid = String.fromCharCode(97 + i);
        }
    }

    if (odd > 1)
        return "";

    const half = cnt.map(x => Math.floor(x / 2));

    const halfLen = Math.floor(n / 2);

    let left = "";

    function build(leftPart) {

        let res = leftPart;

        if (mid !== '#')
            res += mid;

        for (let i = leftPart.length - 1; i >= 0; i--) {
            res += leftPart[i];
        }

        return res;
    }

    for (let pos = 0; pos < halfLen; pos++) {

        let found = false;

        for (let c = 0; c < 26; c++) {

            if (half[c] === 0)
                continue;

            half[c]--;
            left += String.fromCharCode(97 + c);

            // Largest possible completion
            let maxLeft = left;

            for (let x = 25; x >= 0; x--) {
                maxLeft += String.fromCharCode(97 + x)
                    .repeat(half[x]);
            }

            const candidate = build(maxLeft);

            if (candidate > target) {
                found = true;
                break;
            }

            half[c]++;
            left = left.slice(0, -1);
        }

        if (!found)
            return "";
    }

    const ans = build(left);

    return ans > target ? ans : "";
};